### PR TITLE
Feature/expr eval shape

### PR DESCRIPTION
**Pull request by vincentchabannes**
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/feelpp/feelpp/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes?
- [x] Have you successfully run the Feel++ testsuite with your changes locally?
- [ ] Have you written Doxygen comments in your contribution ?

-----

- Allow evaluating a feel++ expression without geometric/finite context (only possible if the expression doesn't depends of a geo/fe context) 
- evaluate(...) return Eigen::Matrix
- fix compilation with various kind of symbols expression

### CODE REVIEW

**Comment by vincentchabannes** [comment]
@prudhomm if you have time, can you do the review

File: feelpp/feel/feelvf/detail/ginacmatrix.hpp [hunk 1]
```diff
@@ -650,14 +671,31 @@ private :
                         {
                             for ( auto const& e : evec )
                             {
-                                uint16_type idx = this->index( e.first );
-                                if ( idx == invalid_uint16_type_value )
-                                    continue;
-                                auto const& theexpr = e.second;
-                                if ( theexpr.isPolynomial() )
-                                    symbTotalDegree.push_back( std::make_pair( M_syms[idx], theexpr.polynomialOrder() ) );
+                                auto const& theexpr = std::get<1>( e );
```
Comment (prudhomm) [inline feelpp/feel/feelvf/detail/ginacmatrix.hpp]: how about using structured bindings to ease reading the code?

File: feelpp/feel/feelvf/detail/ginacmatrix.hpp [hunk 2]
```diff
@@ -703,11 +741,27 @@ private :
                         {
                             for ( auto const& e : evec )
                             {
-                                uint16_type idx = this->index( e.first );
-                                if ( idx == invalid_uint16_type_value )
-                                    continue;
-                                auto const& theexpr = e.second;
-                                x[idx] = theexpr.evaluate( parallel, worldcomm );
+                                auto const& theexpr = std::get<1>( e );
```
Comment (prudhomm) [inline feelpp/feel/feelvf/detail/ginacmatrix.hpp]: use structure bindings ?

File: feelpp/feel/feelvf/matvec.hpp [hunk 3]
```diff
@@ -820,6 +842,13 @@ class Vec
     //! expression is polynomial?
     bool isPolynomial() const { return fusion::accumulate( M_expr, true, GetIsPolynomial{} ); }
 
+    //! evaluate the expression without context
+    evaluate_type evaluate( bool, worldcomm_ptr_t const& ) const
+        {
+            evaluate_type res;
+            fusion::for_each( M_expr,vf::detail::evaluate_expression<evaluate_type>( res ) );
```
Comment (prudhomm) [inline feelpp/feel/feelvf/matvec.hpp]: how about using hana? that would simplify the code, make it more readable and avoid the class and it compiles faster according the hana's benchmarks

File: feelpp/feel/feelvf/matvec.hpp [hunk 4]
```diff
@@ -1120,6 +1149,14 @@ class Mat
...
```
Comment (prudhomm) [inline feelpp/feel/feelvf/matvec.hpp]: see above

File: feelpp/feel/feelvf/stdmathfunctors.hpp [hunk 5]
```diff
@@ -392,6 +402,18 @@ sign( T const& x )
                                                                         \
         bool isPolynomial() const { return  ( M_expr_1.isPolynomial() && M_expr_2.isPolynomial() && ( M_expr_1.polynomialOrder() == 0 ) && ( M_expr_2.polynomialOrder() == 0 ) ); } \
                                                                         \
+        evaluate_type evaluate( bool p, worldcomm_ptr_t const& worldcomm ) const \
+        {                                                               \
+            auto eval1 = M_expr_1.evaluate(p,worldcomm);                \
+            auto eval2 = M_expr_2.evaluate(p,worldcomm);                \
+            CHECK( eval1.rows() == eval2.rows() && eval1.cols() == eval2.cols() ) << "should be have same shape"; \
```
Comment (prudhomm) [inline feelpp/feel/feelvf/stdmathfunctors.hpp]: use DCHECK instead ?

**Pull request merged**